tests/modules/encoder_base_test.py

```diff
@@ -0,0 +1,161 @@
+# pylint: disable=no-self-use,protected-access
+import numpy
+import torch
+from torch.autograd import Variable
+from torch.nn import LSTM
+
+from allennlp.modules.encoder_base import _EncoderBase
+from allennlp.common.testing import AllenNlpTestCase
+from allennlp.nn.util import sort_batch_by_length, get_lengths_from_binary_sequence_mask
+
+
+class TestEncoderBase(AllenNlpTestCase):
+
+    def setUp(self):
+        super(TestEncoderBase, self).setUp()
+        self.lstm = LSTM(bidirectional=True, num_layers=3, input_size=3, hidden_size=7, batch_first=True)
+        self.encoder_base = _EncoderBase(stateful=True)
+
+        tensor = Variable(torch.rand([5, 7, 3]))
+        tensor[1, 6:, :] = 0
+        tensor[2, :, :] = 0  # <= completely masked
+        tensor[3, 2:, :] = 0
+        tensor[4, :, :] = 0  # <= completely masked
+        self.tensor = tensor
+        mask = Variable(torch.ones(5, 7))
+        mask[1, 6:] = 0
+        mask[2, :] = 0
+        mask[3, 2:] = 0
+        mask[4, :] = 0
+        self.mask = mask
+
+        self.batch_size = 5
+        self.num_valid = 4
+        sequence_lengths = get_lengths_from_binary_sequence_mask(mask)
+        _, _, restoration_indices, sorting_indices = sort_batch_by_length(tensor, sequence_lengths)
+        self.sorting_indices = sorting_indices
+        self.restoration_indices = restoration_indices
+
+    def test_get_initial_states(self):
+        # First time we call it, there should be no state, so we should return None.
+        assert self.encoder_base._get_initial_states(self.batch_size, self.num_valid, self.sorting_indices) is None
+
+        # First test the case that the previous state is _smaller_ than the current state input.
+        self.encoder_base._states = (Variable(torch.ones([1, 3, 7])), Variable(torch.ones([1, 3, 7])))
+        # sorting indices are: [0, 1, 3, 2, 4]
+        returned_states = self.encoder_base._get_initial_states(self.batch_size,
+                                                                self.num_valid,
+                                                                self.sorting_indices)
+
+        correct_expanded_state = torch.cat([torch.ones([1, 3, 7]), torch.zeros([1, 2, 7])], 1)
+        # State should have been expanded with zeros to have shape (1, batch_size, hidden_size).
+        numpy.testing.assert_array_equal(self.encoder_base._states[0].data.numpy(),
+                                         correct_expanded_state.numpy())
+        numpy.testing.assert_array_equal(self.encoder_base._states[1].data.numpy(),
+                                         correct_expanded_state.numpy())
+
+        # The returned states should be of shape (1, num_valid, hidden_size) and they also should
+        # have been sorted with respect to the indices.
+        # sorting indices are: [0, 1, 3, 2, 4]
+        correct_returned_state = torch.ones([1, 4, 7])
+        correct_returned_state[:, 2, :] = 0
```

Comment:
????

Reply:
index 2 is completely masked, so the final states are the initial states == 0

Reply:
But this is supposed to only have `num_valid` returned states, and so index 2 shouldn't even be in here...  I'm really confused about how `num_valid` is actually computed, because it doesn't look like it matches what the docs say above.

Reply:
Sorry, this was confusing and only passing because I was using `ones` for the returned state - I have switched it to using random states like the other test (which was passing because it doesn't use `num_valid` at all). My bad.